src/main/scala/io/iohk/ethereum/network/PeerStatisticsActor.scala

```diff
@@ -0,0 +1,115 @@
+package io.iohk.ethereum.network
+
+import akka.actor._
+import cats._
+import cats.implicits._
+import io.iohk.ethereum.network.PeerEventBusActor._
+import io.iohk.ethereum.network.p2p.messages.Codes
+import java.time.Clock
+import scala.concurrent.duration.FiniteDuration
+import cats.kernel.Monoid
+
+class PeerStatisticsActor(
+    peerEventBus: ActorRef,
+    var maybeStats: Option[TimeSlotStats[PeerId, PeerStatisticsActor.Stat]]
+) extends Actor {
+  import PeerStatisticsActor._
+
+  override def preStart(): Unit = {
+    // Subscribe to messages received from handshaked peers to maintain stats.
+    peerEventBus ! Subscribe(MessageSubscriptionClassifier)
+    // Removing peers is an optimisation to free space, but eventually the stats would be overwritten anyway.
+    peerEventBus ! Subscribe(SubscriptionClassifier.PeerDisconnectedClassifier(PeerSelector.AllPeers))
+  }
+
+  def receive: Receive = handlePeerEvents orElse handleStatsRequests
+
+  private def handlePeerEvents: Receive = {
+    case PeerEvent.MessageFromPeer(msg, peerId) =>
+      val now = System.currentTimeMillis()
+      val obs = Stat(
+        responsesReceived = if (ResponseCodes(msg.code)) 1 else 0,
+        requestsReceived = if (RequestCodes(msg.code)) 1 else 0,
+        firstSeenTimeMillis = Some(now),
+        lastSeenTimeMillis = Some(now)
+      )
+      maybeStats = maybeStats.map(_.add(peerId, obs))
+
+    case PeerEvent.PeerDisconnected(peerId) =>
+      maybeStats = maybeStats.map(_.remove(peerId))
+  }
+
+  private def handleStatsRequests: Receive = {
+    case GetStatsForAll(window) =>
+      val stats = maybeStats.map(_.getAll(Some(window))).getOrElse(Map.empty)
+      sender ! StatsForAll(stats)
+
+    case GetStatsForPeer(window, peerId) =>
+      val stats = maybeStats.map(_.get(peerId, Some(window))).getOrElse(Stat.empty)
+      sender ! StatsForPeer(peerId, stats)
+  }
+}
+
+object PeerStatisticsActor {
+  case class Stat(
```

Comment:
Maybe `Stat` is big enough to live in its own file?

Reply:
Done.